File: src/main/java/com/learn/todolist/infra/JwtAuthorizationFilter.java
package com.learn.todolist.infra;

import com.learn.todolist.config.AllowedConfig;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.Arrays;

@RequiredArgsConstructor
@Component
public class JwtAuthorizationFilter extends OncePerRequestFilter {
    private final TokenProvider tokenProvider;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        String token = request.getHeader("Authorization");
        if (token != null) {
            token = token.replace("Bearer", "").trim();
        } else {
            token = "";
        }
        if (Arrays.asList(AllowedConfig.ALLOWED_URLS).contains(request.getRequestURI())) {
            filterChain.doFilter(request, response);
            return;
        }

        if (!token.isBlank()) {
            if (!tokenProvider.validateToken(token)) {
                throw new IllegalArgumentException("Invalid Token");
            }
            SecurityContextHolder.getContext().setAuthentication(tokenProvider.authenticateToken(token));
            filterChain.doFilter(request, response);
            return;
        }

        filterChain.doFilter(request, response);
    }
}

File: src/main/kotlin/com/todolist/infra/security/JWTAuthorizationFilter.kt
package com.todolist.infra.security

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter


@Component
class JWTAuthorizationFilter(
    private val tokenProvider: TokenProvider
): OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        var token = ""
        request.getHeader("Authorization")?.let {
            token = it.replace("bearer", "", ignoreCase = true).trimIndent()
        }
        when(request.requestURI){
            "/api/v1/auth/login", "/api/v1/registers/register", "/swagger-ui/**", "/v3/api-docs" -> {
                filterChain.doFilter(request, response)
                return
            }
            else -> {
                if (token.isNotBlank()){
                    check(tokenProvider.validateToken(token)) {"Invalid Token"}
                    SecurityContextHolder.getContext().authentication = tokenProvider.authenticateToken(token)
                }
            }
        }
        filterChain.doFilter(request, response)
        return
    }
}